src/python/zquantum/core/decompositions/_decomposition.py

```diff
@@ -0,0 +1,53 @@
+from abc import abstractmethod
+from typing import Iterable, Sequence, TypeVar
+
+from typing_extensions import Protocol
+
+OperationType = TypeVar("OperationType")
+
+
+class DecompositionRule(Protocol[OperationType]):
+    @abstractmethod
+    def production(self, operation: OperationType) -> Iterable[OperationType]:
+        """Decompose given operation.
+
+        Implementers should make sure that the result of applying the decomposed
+        sequence to any wavefunction is the same as after applying the original
+        operation.
+        """
+
+    @abstractmethod
+    def predicate(self, operation: OperationType) -> bool:
+        """Determine if this rule should be applied to given operation."""
+
+
+def decompose_operation(
+    operation: Iterable[OperationType],
+    decomposition_rules: Sequence[DecompositionRule[OperationType]],
+):
+    if not decomposition_rules:
+        return [operation]
+
+    current_rule, remaining_rules = decomposition_rules[0], decomposition_rules[1:]
```

Comment:
FYI, there's a destructuring syntax for this exact purpose :)
```suggestion
    current_rule, *remaining_rules = decomposition_rules
```

Reply:
I thought it is less known, but on the other hand it's much cleaner. Done in  e69e23e